python/mlc_chat/compiler/compile.py

```diff
@@ -89,65 +70,60 @@ def _attach_preproc(
+        if hasattr(model_config, "sliding_window_size"):
+            return {
+                "seq_len": model_config.prefill_chunk_size,
+                "rolling_cache_len": model_config.sliding_window_size,
+                "kv_seq_len": model_config.sliding_window_size + model_config.prefill_chunk_size,
+            }
+        return {
+            "seq_len": model_config.prefill_chunk_size,
+            "total_seq_len": model_config.context_window_size,
         }
-    )
 
+    def _get_param_metadata(name: str, param: nn.Parameter) -> Dict[str, Any]:
+        return {
+            "name": name,
+            "shape": list(param.shape),
+            "dtype": param.dtype,
+            "preprocs": param.attrs["preprocs"],
+        }
 
-def _compile(args: CompileArgs, model_config: ConfigBase):
     logger.info("Creating model from: %s", args.config)
     args.overrides.apply(model_config)
     with args.target:
         # Step 1. Create the quantized model
         model, _ = args.model.quantize[args.quantization.kind](model_config, args.quantization)
         # Step 2. Exporting the model to TVM Unity
         logger.info("Exporting the model to TVM Unity compiler")
-        mod, named_params = model.export_tvm(
-            spec=model.get_default_spec(),  # type: ignore
-        )
-        _attach_variable_bounds(mod, model_config)
-        _attach_preproc(mod, named_params, args, model_config)
+        mod, named_params = model.export_tvm(spec=model.get_default_spec())  # type: ignore
         # Step 3. Running relax compilation pipeline
         logger.info("Running optimizations using TVM Unity")
-        mod = relax.get_pipeline("mlc_llm")(mod)
-        # Step 4. Attach metadata for the runtime to read
-        _attach_metadata(mod, named_params, args, model_config)
-        logger.info("Generating code using TVM Unity")
-        # Step 5. Build and export the library
-        args.build_func(mod, args)
+        additional_tirs = _apply_preproc_to_params(named_params, model_config)
+        variable_bounds = _get_variable_bounds(model_config)
+        args.build_func(
+            mod,
+            args,
+            pipeline=relax.get_pipeline(
+                "mlc_llm",
+                variable_bounds=variable_bounds,
+                additional_tirs=additional_tirs,
+                metadata={
+                    "model_type": args.model.name,
+                    "quantization": args.quantization.name,
+                    "params": [_get_param_metadata(name, param) for name, param in named_params],
+                    "context_window_size": model_config.context_window_size,  # type: ignore
+                    "prefill_chunk_size": model_config.prefill_chunk_size,  # type: ignore
+                    "sliding_window_size": getattr(model_config, "sliding_window_size", -1),
+                    "tensor_parallel_shards": model_config.tensor_parallel_shards,  # type: ignore
+                },
+            ),
+        )
```

Comment:
These lines demonstrates the effectiveness of https://github.com/apache/tvm/pull/16246 - by specifying the `pipeline` parameter in `relax.build`, now it runs with arbitrary compilation passes we specify.